### DISCUSSION TITLE
using the vLLM instead of ollama

### DISCUSSION THREAD

**Discussion by iukea1** [open-webui/open-webui#508] [Q&A] [answered]
Has anyone tried using vLLM to run inference on their models and serve them via this WebUI. Just curious how much of a can of worms that is .

Is it as simple as just pointing  to the VLLMs URL 🤔 

Reason for asking I was interested in AWQ and using my 4090 more efficiently if possible.

**Comment by justinh-rahb** [comment]
Greetings @iukea1, while "never" might not quite fit here, it's accurate to say that for now, the **Ollama** WebUI project is closely tied with **Ollama**...🦙

Personally I agree that this direction could pique the interest of some individuals. In my view, this potential divergence may be an acceptable reason for a friendly project fork. It's possible that down the line, these branches might merge again if they remain aligned otherwise.

At a high-level, most of the needed changes would be in the backend, specifically the ollama API modules, most of the rest could stay the same. You'd remove the model downloading functionality from the frontend as AFAIK vLLM doesn't have such a capability (correct me if I'm wrong, I don't pay much attention to GPU-related projects as I have no budget to rent/buy any 😂), otherwise again you could leave most of it alone probably. The sister-project [Ollama WebUI Lite](https://github.com/ollama-webui/ollama-webui-lite) would be an ideal starting place for such a project too, since it comprises only the frontend code.

**Comment by tjbck** [comment]
I haven't tested it myself but It should work out the box by [deploying vLLM as an OpenAI-compatible server.](https://docs.vllm.ai/en/latest/getting_started/quickstart.html#openai-compatible-server.) Keep us updated if you manage to get it working!

**Reply by justinh-rahb** [reply, marked as answer]
Quite right @tjbck you could just as easily not modify anything and simply use vLLM as an OpenAI endpoint. It does present an issue if one doesn't have even an empty Ollama instance to connect to, but it shouldn't be hard for someone to mock up an empty Ollama API models endpoint to bypass the check if they were so motivated.

**Reply by aitzubi** [reply]
@tjbck  @justinh-rahb I am trying to make this work and it is not working. Has anyone manage to make this work? I just created a model, pointed to our endpoint (http://vllmhost:port/v1) and tried some models but it is not working. It returns error 500.

Any hints?

**Reply by justinh-rahb** [reply]
> @tjbck  @justinh-rahb I am trying to make this work and it is not working. Has anyone manage to make this work? I just created a model, pointed to our endpoint (http://vllmhost:port/v1) and tried some models but it is not working. It returns error 500.
> 
> Any hints?

Treat it as an `OpenAI compatible` endpoint, don't use the `vllm` format in LiteLLM.

**Reply by aitzubi** [reply]
Yeah, just figured it out. We were not using an api key on our VLLM endpoints so we were specifying an empty string as our API_KEY parameter value. We changed it to API_KEY='Empty' and it started working.

Thank you!

**Comment by flefevre** [comment]
Hello
Could you provide more on how you deal to do it ?
I would like to beable to déployé vllm and make it accessible through thé open webui application which is définitively user friendly and performant. 

Sorry for m'y english, i am french

**Reply by aitzubi** [reply]
Sure!

You have to create a config.yaml for litellm similar to this:

>- litellm_params:
    api_base: http://yourvllmurl:8000/v1
    api_key: yourapikey
    model: openai/google/gemma-7b-it
  model_name: gemma-7b-it
>- litellm_params:
    api_base: http://yourvllmurl:8001/v1
    api_key: yourapikey
    model: openai/TheBloke/Llama-2-70B-Chat-GPTQ
  model_name: Llama-2-70B-Chat-GPTQ 

Then you map that config file in your docker run command like this:

`docker run -d -p 3000:8080 --add-host=host.docker.internal:host-gateway -v open-webui:/app/backend/data -v /home/username/litellm/config.yaml:/app/backend/data/litellm/config.yaml`

And that's it!

**Reply by flefevre** [reply]
Thanks a lot. I will integrate it. Buty just to be sure, how can I test it? to be sure I am going from open webui to litellm and not to ollama? is there a basic command to see it?

**Reply by flefevre** [reply]
I have tried several times but i think there is a problem in my docker-compose file

`
open-webui:
    #extensible, feature-rich, and user-friendly self-hosted WebUI designed to operate entirely offline
    #https://github.com/open-webui/open-webui
    profiles: ["fullstack"]
    image: ghcr.io/open-webui/open-webui:git-2fa9495
    container_name: cnopenwebui
    volumes:
      - ./open-webui/open-webui:/app/backend/data
      #https://docs.openwebui.com/tutorial/litellm
      #- ./litellm/config/config.yaml:/app/backend/data/litellm/config.yaml
    depends_on:
      - ollama
      #- redis
      #- litellm
    ports:
      - ${OPEN_WEBUI_PORT-4000}:8080
    env_file:
      - .env
    environment:
      #- 'OLLAMA_BASE_URL=http://ollama:11434'
      - 'OLLAMA_BASE_URLS=http://ollama:11434;http://ollamacpu:11435'
      - 'WEBUI_SECRET_KEY='
      #- 'OPENAI_API_BASE_URLS=${OPENAI_API_BASE_URLS}'
      #- 'OPENAI_API_KEYS=${OPENAI_API_KEYS}'
      #- 'REDIS_HOST=redis'
    extra_hosts:
      - host.docker.internal:host-gateway
    restart: unless-stopped
    networks:
      - netiagen4spring

redis:
    #cache configuration
    profiles: ["fullstack"]
    image: redis:7.2.4-alpine
    container_name: cnredis
    command:
      - sh
      - -c # this is to evaluate the $REDIS_PASSWORD from the env
      - redis-server --appendonly yes --requirepass $$REDIS_PASSWORD ## $$ because of docker-compose
    env_file:
      - .env
    environment:
      REDIS_PASSWORD: ${REDIS_PASSWORD}
    volumes:
      - ./redis/data:/data
    networks:
      - netiagen4spring

  litellm:
    #https://github.com/BerriAI/litellm/
    profiles: ["fullstack"]
    image: ghcr.io/berriai/litellm:main-v1.30.2
    container_name: cnlitellm
    volumes:
      - ./litellm/config/config.yaml:/app/config.yaml
    command:
      [
        "--config",
        "/app/config.yaml",
        "--port",
        "8001",
        "--num_workers",
        "8"
      ]
    depends_on:
      - ollama
      - redis
    ports:
      - 8001:8001
    env_file:
      - .env
    environment:
      REDIS_HOST: redis
      REDIS_PORT: ${REDIS_PORT}
      REDIS_PASSWORD: ${REDIS_PASSWORD}
      UI_USERNAME: ${UI_USERNAME}
      UI_PASSWORD: ${UI_PASSWORD}
      LITELLM_LOG: DEBUG #INFO,NONE
      DATABASE_URL: postgresql://postgres:postgres@db:5432/postgres
    networks:
      - netiagen4spring

`

When i start i get the following errors

`cnopenwebui    | Setting Cache on Proxy
cnopenwebui    | Cache Type: redis
cnopenwebui    | Cache Host: redis
cnopenwebui    | Cache Port: 6379
cnopenwebui    | Cache Password: XXXXXXXXXXXXX
cnopenwebui    | 
cnopenwebui    | INFO:     Started server process [1]
cnopenwebui    | INFO:     Waiting for application startup.
cnopenwebui    | ERROR:    Traceback (most recent call last):
cnopenwebui    |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 734, in lifespan
cnopenwebui    |     async with self.lifespan_context(app) as maybe_state:
cnopenwebui    |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 610, in __aenter__
cnopenwebui    |     await self._router.startup()
cnopenwebui    |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 711, in startup
cnopenwebui    |     await handler()
cnopenwebui    |   File "/app/backend/main.py", line 144, in on_startup
cnopenwebui    |     await litellm_app_startup()
cnopenwebui    |   File "/app/backend/apps/litellm/main.py", line 33, in startup
cnopenwebui    |     await config()
cnopenwebui    |   File "/app/backend/apps/litellm/main.py", line 25, in config
cnopenwebui    |     router, model_list, general_settings = await proxy_config.load_config(
cnopenwebui    |                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cnopenwebui    |   File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 1584, in load_config
cnopenwebui    |     litellm.cache = Cache(**cache_params)
cnopenwebui    |                     ^^^^^^^^^^^^^^^^^^^^^
cnopenwebui    |   File "/usr/local/lib/python3.11/site-packages/litellm/caching.py", line 805, in __init__
cnopenwebui    |     self.cache: BaseCache = RedisCache(host, port, password, **kwargs)
cnopenwebui    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cnopenwebui    |   File "/usr/local/lib/python3.11/site-packages/litellm/caching.py", line 98, in __init__
cnopenwebui    |     from ._redis import get_redis_client, get_redis_connection_pool
cnopenwebui    |   File "/usr/local/lib/python3.11/site-packages/litellm/_redis.py", line 13, in <module>
cnopenwebui    |     import redis, litellm
cnopenwebui    | ModuleNotFoundError: No module named 'redis'
cnopenwebui    | 
cnopenwebui    | ERROR:    Application startup failed. Exiting.
cnopenwebui    | 
cnopenwebui    | Setting Cache on Proxy
cnopenwebui    | Cache Type: redis
cnopenwebui    | Cache Host: redis
cnopenwebui    | Cache Port: 6379
cnopenwebui    | Cache Password: SDsdgFsfm4lmf4SSWORD}
cnopenwebui    | 
`

it doesn't find the redis module ? used by litellm for cache...

**Comment by G4Zz0L1** [comment]
I'm having the same error.
Everything on latest (open-webui and litellm).
Without redis, everything works fine, but I lose the ability to load balance with litellm.
If I add the redis container for litellm, open-webui gives the error below.
Any idea what to fix, or who to blame between the two?

BTW, @flefevre, format your issue correctly, I don't think it's easy to understand what you've pasted there.

```
open-webui  | INFO:     Started server process [1]
open-webui  | INFO:     Waiting for application startup.
open-webui  | ERROR:    Traceback (most recent call last):
open-webui  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 732, in lifespan
open-webui  |     async with self.lifespan_context(app) as maybe_state:
open-webui  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 608, in __aenter__
open-webui  |     await self._router.startup()
open-webui  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 709, in startup
open-webui  |     await handler()
open-webui  |   File "/app/backend/main.py", line 165, in on_startup
open-webui  |     await litellm_app_startup()
open-webui  |   File "/app/backend/apps/litellm/main.py", line 38, in startup
open-webui  |     await config()
open-webui  |   File "/app/backend/apps/litellm/main.py", line 30, in config
open-webui  |     router, model_list, general_settings = await proxy_config.load_config(
open-webui  |                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
open-webui  |   File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 1584, in load_config
open-webui  |     litellm.cache = Cache(**cache_params)
open-webui  |                     ^^^^^^^^^^^^^^^^^^^^^
open-webui  |   File "/usr/local/lib/python3.11/site-packages/litellm/caching.py", line 805, in __init__
open-webui  |     self.cache: BaseCache = RedisCache(host, port, password, **kwargs)
open-webui  |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
open-webui  |   File "/usr/local/lib/python3.11/site-packages/litellm/caching.py", line 98, in __init__
open-webui  |     from ._redis import get_redis_client, get_redis_connection_pool
open-webui  |   File "/usr/local/lib/python3.11/site-packages/litellm/_redis.py", line 13, in <module>
open-webui  |     import redis, litellm
open-webui  | ModuleNotFoundError: No module named 'redis'
open-webui  | 
open-webui  | ERROR:    Application startup failed. Exiting.
```

**Comment by justinh-rahb** [comment]
@flefevre @G4Zz0L1, It looks like there is a misunderstanding with how we utilize LiteLLM internally in our project. The crux of the problem lies in an attempt to use a single configuration file for both the internal LiteLLM instance embedded within Open WebUI and the separate, external LiteLLM container that has been added. This approach has led to an untenable situation, especially regarding the Redis configuration settings.

Our internal LiteLLM setup, as designed for use within Open WebUI, does not support Redis caching. When the external LiteLLM container includes Redis settings within the shared configuration file, it creates a conflict. The internal LiteLLM instance, upon encountering Redis settings, cannot process them, as it's not designed to interact with Redis.

**Reply by G4Zz0L1** [reply]
You're right, I've just tried with two differente config.yaml, and everything seems to work.
Thanks a lot for this one!
Now the only block it's the fact that this config its unmanegeable from the open-webui part, because it's not reflected in the litellm one.
Do you have any ideas on how to overcome this?

**Reply by justinh-rahb** [reply]
> You're right, I've just tried with two differente config.yaml, and everything seems to work.
> Thanks a lot for this one!
> Now the only block it's the fact that this config its unmanegeable from the open-webui part, because it's not reflected in the litellm one.
> Do you have any ideas on how to overcome this?

Unfortunately no, our UI was only designed with using the internal instance in mind. It was assumed that if one wanted to use more LiteLLM features than we support that they'd run it as it's own container with a handcrafted config file.

**Reply by G4Zz0L1** [reply]
I see, so there's no way to sync even only the "model_list" key, without using an external agent or script to do it.
Ok, thanks for the explanation.

**Reply by justinh-rahb** [reply]
> I see, so there's no way to sync even only the "model_list" key, without using an external agent or script to do it. Ok, thanks for the explanation.

For the time being, yes, something like that would be necessary. Not that we haven't ruled out having an architecture whereby one could use _either_ internal or external instance (and for other things too such as databases, or RAG for example) based on their requirement for scale.

**Comment by gaby** [comment]
Was anyone able to do this and can provide a docker-compose.yml example to run both tools together?

**Comment by uocnb** [comment]
Here is my docker compose for Local LLM setup, including:
1. vllm
2. ollama
3. open-webui
4. open-webui-pipelines

You can get it [here](https://gist.github.com/uocnb/ed6b6764c8005aabc78a37ff027ec3ee)
Have fun!

**Reply by NabuchodonosorII** [reply]
No mention to chat template at all ?

**Reply by NabuchodonosorII** [reply]
Ok Vllm as Default template in the tokeniser, but an other question, how do we supposed to access to Open WebUI interface, i tried to use the http://localhost:8089, but I have nothing. Here is my docker compose :

```yaml
version: '3'
services:
  vllm:
    image: vllm/vllm-openai:latest
    container_name: vllm
    hostname: vllm
    runtime: nvidia
    deploy:
        resources:
            reservations:
                devices:
                    - driver: nvidia 
                      count: 1 
                      capabilities: [gpu]  
    
    ipc: host
    environment:
      - HUGGING_FACE_HUB_TOKEN=my hugging face token
      - VLLM_CPU_KVCACHE_SPACE=64
    command:
      - "--device"
      - "auto"
      - "--max-num-seqs"
      - "2"
      - "--kv-cache-dtype"
      - "fp8"
      - "--load-format"
      - "gguf"
      - "--model"
      - "/models/bartowski/Qwen2.5-32B-Instruct-GGUF/Qwen2.5-32B-Instruct-Q3_K_M.gguf"
     
    ports:
      - "8042:8000/tcp"
    restart: always
    volumes:
      - ~/.cache/huggingface:/root/.cache/huggingface
      - ./data/.ollama:/root/.ollama
      - H:/LLM:/models 

  ollama:
    image: ollama/ollama
    container_name: ollama
    hostname: ollama
    environment:
      - OLLAMA_HOST=0.0.0.0
    ipc: host
    ports:
      - "11434:11434/tcp"
    restart: always
    volumes:
      - ./data/.ollama:/root/.ollama
      - H:/LLM:/models 

  open-webui:
    image: ghcr.io/open-webui/open-webui:main
    container_name: open-webui
    hostname: open-webui
    environment:
      - OLLAMA_PORT=11434
      - OLLAMA_HOST=127.0.0.1
      - OLLAMA_BASE_URL=http://127.0.0.1:11434
      - OLLAMA_BASE_URLS=http://127.0.0.1:11434
      - PORT=8089  
    extra_hosts:
      - host.docker.internal:host-gateway
    network_mode: host 
    restart: always
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - /etc/timezone:/etc/timezone:ro
      - ./data/open-webui:/app/backend/data

  open-webui-pipelines:
    image: ghcr.io/open-webui/pipelines:main
    container_name: open-webui-pipelines
    hostname: open-webui-pipelines
    extra_hosts:
      - host.docker.internal:host-gateway
    network_mode: host
    restart: always
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - /etc/timezone:/etc/timezone:ro
      - ./data/open-webui-pipelines:/app/pipelin

**Reply by uocnb** [reply]
> No mention to chat template at all ?

Chat template can be configured on the open-webui.

> Ok Vllm as Default template in the tokeniser, but an other question, how do we supposed to access to Open WebUI interface, i tried to use the http://localhost:8089/, but I have nothing. Here is my docker compose :

Since it using the host network mode, all exposed ports from the container can be access. Also try with `http://127.0.0.1:8089/`, I have issue with IPv6 sometime that make localhost not accessible.
You might want to check with `docker compose ps` to see if the open-webui up or not. Check the log with `docker compose logs open-webui`.

**Reply by NabuchodonosorII** [reply]
`curl:`` (7) Failed to connect to 127.0.0.1 port 8089 after 2016 ms: Could not connect to server

Dont know why it is disconnected, but the open webui is up 
![image](https://github.com/user-attachments/assets/1b80d750-f49d-4457-84f3-46b7223c4abe)

So I'm a bit confused, maybe it's the things I did modify from your docker compose, can you see something that should not be there in my docker compose ?

**Reply by AboveTheClouds90** [reply]
> Here is my docker compose for Local LLM setup, including:
> 
>     1. vllm
> 
>     2. ollama
> 
>     3. open-webui
> 
>     4. open-webui-pipelines
> 
> 
> You can get it [here](https://gist.github.com/uocnb/ed6b6764c8005aabc78a37ff027ec3ee) Have fun!

Hello,
Im relatively new and wanted to build a setup for a h100 so multiple users can access it. Why exactly did you include ollama in this setup? Shouldnt vllm be enough for handling inference/backend and batching? So why do you need ollama for?
